archive/schema.go

```diff
@@ -107,25 +81,104 @@ func (c *CreateOptions) toMetadata() *Metadata {
 }
 
 type Archive struct {
-	Meta *Metadata
-	Root string
-
-	// Spans contains either all spans from metadata, or a subset
-	// due to opening the archive with a filter list.
-	Spans []SpanInfo
+	Root              string
+	DataSortDirection zbuf.Direction
+	LogSizeThreshold  int64
+	LogsFiltered      bool
+
+	// mu protects below fields.
+	mu    sync.RWMutex
+	spans []SpanInfo
+	// mdModTime is the mtime of the metadata at or before its contents
+	// were last read.
+	mdModTime time.Time
+	// mdUpdateCount is incremented every time the metadata for the
+	// archive is potentially updated due to writing new logs, or
+	// on re-reading the metadata file.
+	mdUpdateCount int
 }
 
 func (ark *Archive) AppendSpans(spans []SpanInfo) error {
-	ark.Meta.Spans = append(ark.Meta.Spans, spans...)
+	if ark.LogsFiltered {
+		return errors.New("cannot add spans to log filtered archive")
+	}
+
+	ark.mu.Lock()
+	defer ark.mu.Unlock()
 
-	sort.Slice(ark.Meta.Spans, func(i, j int) bool {
-		if ark.Meta.DataSortDirection == zbuf.DirTimeForward {
-			return ark.Meta.Spans[i].Span.Ts < ark.Meta.Spans[j].Span.Ts
+	ark.spans = append(ark.spans, spans...)
+
+	sort.Slice(ark.spans, func(i, j int) bool {
+		if ark.DataSortDirection == zbuf.DirTimeForward {
+			return ark.spans[i].Span.Ts < ark.spans[j].Span.Ts
 		}
-		return ark.Meta.Spans[j].Span.Ts < ark.Meta.Spans[i].Span.Ts
+		return ark.spans[j].Span.Ts < ark.spans[i].Span.Ts
 	})
 
-	return ark.Meta.Write(filepath.Join(ark.Root, metadataFilename))
+	err := ark.metaWrite()
+	if err != nil {
+		return err
+	}
+
+	ark.mdUpdateCount++
+	return nil
+}
+
+func (ark *Archive) metaWrite() error {
+	m := &Metadata{
+		Version:           0,
+		LogSizeThreshold:  ark.LogSizeThreshold,
+		DataSortDirection: ark.DataSortDirection,
+		Spans:             ark.spans,
+	}
+	return m.Write(ark.mdPath())
+}
+
+func (ark *Archive) mdPath() string {
+	return filepath.Join(ark.Root, metadataFilename)
+}
+
+// UpdateCheck looks at the archive's metadata file to see if it
+// has been written to since last read; if so, it is read and the
+// available spans are updated. A counter is returned, starting
+// from 1, which is incremented every time this Archive has re-read
+// the metadata file, and possibly updated the available spans.
+func (ark *Archive) UpdateCheck() (int, error) {
```

Comment:
Why not just return mdModTime? What does having the counter (and thus an extra piece of state) give us?

Reply:
Oh, I'm guessing this for testing purposes?

Reply:
I did want a counter to aid with testing. I also thought it was a better api to return an integer instead of a date, as the integer was simpler to interpret ('if its not the same as last time, revoke your cache').